pkg/gofr/datasource/opentsdb/put.go

```diff
@@ -0,0 +1,354 @@
+
+	err := validateDataPoint(datas)
+	if err != nil {
+		message = fmt.Sprintf("invalid data: %s", err)
+		return nil, err
+	}
+
+	if !isValidPutParam(queryParam) {
+		message = "The given query param is invalid."
+		return nil, errors.New(message)
+	}
+
+	var putEndpoint string
+	if !isEmptyPutParam(queryParam) {
+		putEndpoint = fmt.Sprintf("%s%s?%s", c.tsdbEndpoint, PutPath, queryParam)
+	} else {
+		putEndpoint = fmt.Sprintf("%s%s", c.tsdbEndpoint, PutPath)
+	}
+
+	dataGroups, err := c.splitProperGroups(datas)
+	if err != nil {
+		message = fmt.Sprintf("split data point error: %s", err)
+		return nil, err
+	}
+
+	responses := make([]PutResponse, 0)
+
+	responses, err = c.getResponses(putEndpoint, dataGroups, responses, &message)
+	if err != nil {
+		return nil, err
+	}
+
+	globalResp := PutResponse{logger: c.logger, tracer: c.tracer, ctx: c.ctx}
+	globalResp.StatusCode = http.StatusOK
+
+	for _, resp := range responses {
+		globalResp.Failed += resp.Failed
+		globalResp.Success += resp.Success
+		globalResp.Errors = append(globalResp.Errors, resp.Errors...)
+
+		if resp.StatusCode != http.StatusOK && globalResp.StatusCode == http.StatusOK {
+			globalResp.StatusCode = resp.StatusCode
+		}
+	}
+
+	if globalResp.StatusCode == http.StatusOK {
+		status = StatusSuccess
+		message = fmt.Sprintf("Put request to url %q processed successfully", putEndpoint)
+
+		return &globalResp, nil
+	}
+
+	return nil, parsePutErrorMsg(&globalResp)
+}
+
+func (c *Client) getResponses(putEndpoint string, dataGroups [][]DataPoint,
+	responses []PutResponse, message *string) ([]PutResponse, error) {
+	for _, datapoints := range dataGroups {
+		reqBodyCnt, err := getPutBodyContents(datapoints)
+		if err != nil {
+			*message = fmt.Sprintf("getPutBodyContents error: %s", err)
+			c.logger.Errorf(*message)
+		}
+
+		putResp := PutResponse{logger: c.logger, tracer: c.tracer, ctx: c.ctx}
+
+		if err = c.sendRequest(PostMethod, putEndpoint, reqBodyCnt, &putResp); err != nil {
+			*message = fmt.Sprintf("error processing put request at url %q: %s", putEndpoint, err)
+			return nil, err
+		}
+
+		responses = append(responses, putResp)
+	}
+
+	return responses, nil
+}
+
+// splitProperGroups splits the given datapoints into groups, whose content size is not larger than
+// c.opentsdbCfg.MaxContentLength. This method is to avoid Put failure, when the content length of
+// the given datapoints in a single /api/put request exceeded the value of
+// tsd.http.request.max_chunk in the opentsdb config file.
+func (c *Client) splitProperGroups(datapoints []DataPoint) ([][]DataPoint, error) {
+	span := c.addTrace(c.ctx, "splitProperGroups-Put")
+
+	status := StatusFailed
+
+	var message string
+
+	defer sendOperationStats(c.logger, time.Now(), "splitProperGroups-Put", &status, &message, span)
+
+	datasBytes, err := json.Marshal(&datapoints)
+	if err != nil {
+		message = fmt.Sprintf("failed to marshal the datapoints to be put: %v", err)
```

Comment:
```suggestion
		message = fmt.Sprintf("failed to marshal the datapoints to be put: %w", err)
```

Reply:
%w is used with fmt.Errorf() calls .....